tests/nexus_workflow_test.go

```diff
@@ -732,6 +733,225 @@ func (s *NexusWorkflowTestSuite) TestNexusOperationAsyncCompletion() {
+	ctx, cancel := context.WithCancel(testcore.NewContext())
+	defer cancel()
+	taskQueue := testcore.RandomizeStr(s.T().Name())
+	endpointName := testcore.RandomizedNexusEndpoint(s.T().Name())
+
+	_, err := s.OperatorClient().CreateNexusEndpoint(ctx, &operatorservice.CreateNexusEndpointRequest{
+		Spec: &nexuspb.EndpointSpec{
+			Name: endpointName,
+			Target: &nexuspb.EndpointTarget{
+				Variant: &nexuspb.EndpointTarget_Worker_{
+					Worker: &nexuspb.EndpointTarget_Worker{
+						Namespace: s.Namespace(),
+						TaskQueue: taskQueue,
+					},
+				},
+			},
+		},
+	})
+	s.NoError(err)
+
+	run, err := s.SdkClient().ExecuteWorkflow(ctx, client.StartWorkflowOptions{
+		TaskQueue: taskQueue,
+	}, "workflow")
+	s.NoError(err)
+
+	completionWFType := "completion_wf"
+	completionWFID := testcore.RandomizeStr(s.T().Name())
+	completionWFTaskQueue := testcore.RandomizeStr(s.T().Name())
+	completionWFStartReq := &workflowservice.StartWorkflowExecutionRequest{
+		RequestId:          uuid.NewString(),
+		Namespace:          s.Namespace(),
+		WorkflowId:         completionWFID,
+		WorkflowType:       &commonpb.WorkflowType{Name: completionWFType},
+		TaskQueue:          &taskqueuepb.TaskQueue{Name: completionWFTaskQueue, Kind: enumspb.TASK_QUEUE_KIND_NORMAL},
+		Input:              nil,
+		WorkflowRunTimeout: durationpb.New(100 * time.Second),
+		Identity:           "test",
+	}
+
+	startLink := &commonpb.Link_WorkflowEvent{
+		Namespace:  s.GetNamespaceID(s.Namespace()),
+		WorkflowId: completionWFID,
+		Reference: &commonpb.Link_WorkflowEvent_EventRef{
+			EventRef: &commonpb.Link_WorkflowEvent_EventReference{
+				EventType: enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_STARTED,
+			},
+		},
+	}
+
+	finishStartReqCh := make(chan struct{})
+	startFinishedCh := make(chan struct{})
+	// Special Nexus task poller to handle waiting until start request is finished.
+	go func(t *testing.T, ns string, feClient testcore.FrontendClient) {
+		defer close(startFinishedCh)
+
+		res, err := feClient.PollNexusTaskQueue(ctx, &workflowservice.PollNexusTaskQueueRequest{
+			Namespace: ns,
+			Identity:  uuid.NewString(),
+			TaskQueue: &taskqueuepb.TaskQueue{
+				Name: taskQueue,
+				Kind: enumspb.TASK_QUEUE_KIND_NORMAL,
+			},
+		})
+		require.NoError(t, err)
+
+		start := res.Request.Variant.(*nexuspb.Request_StartOperation).StartOperation
+		require.Equal(t, op.Name(), start.Operation)
+
+		start.CallbackHeader[nexus.HeaderOperationID] = completionWFID
+		completionWFStartReq.CompletionCallbacks = []*commonpb.Callback{
+			{
+				Variant: &commonpb.Callback_Nexus_{
+					Nexus: &commonpb.Callback_Nexus{
+						Url:    start.Callback,
+						Header: start.CallbackHeader,
+					},
+				},
+			},
+		}
+
+		completionRun, err := feClient.StartWorkflowExecution(ctx, completionWFStartReq)
+		require.NoError(t, err)
+		startLink.RunId = completionRun.RunId
+
+		// Delay finishing start request until after completion is received.
+		<-finishStartReqCh
```

Comment:
You don't need this. Just fail the request.

Reply:
I wanted to test that the server handles getting the started response appropriately when they are received out of order. Failing the request means we would never see the started response which doesn't quite match the real-world race condition.